### PR TITLE
Added ported C++ version of ISAM2 Kitti example

### PR DESCRIPTION
**Pull request by mindThomas**
Tested, compared and verified against MATLAB results.

---
This change is [Reviewable](https://reviewable.io/reviews/borglab/gtsam/249)

### CODE REVIEW

File: examples/IMUKittiExampleGPS.cpp [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+#include <gtsam/navigation/ImuFactor.h>
+#include <gtsam/slam/dataset.h>
+#include <gtsam/slam/BetweenFactor.h>
+#include <gtsam/slam/PriorFactor.h>
+#include <gtsam/nonlinear/ISAM2.h>
+#include <gtsam/nonlinear/ISAM2Params.h>
+#include <gtsam/nonlinear/LevenbergMarquardtOptimizer.h>
+#include <gtsam/nonlinear/NonlinearFactorGraph.h>
+#include <gtsam/inference/Symbol.h>
+#include <cstring>
```
Comment (dellaert) [inline examples/IMUKittiExampleGPS.cpp]: Separate group

File: examples/IMUKittiExampleGPS.cpp [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+#include <iostream>
+
+using namespace gtsam;
+using namespace std;
+
+using symbol_shorthand::X; // Pose3 (x,y,z,r,p,y)
+using symbol_shorthand::V; // Vel   (xdot,ydot,zdot)
+using symbol_shorthand::B; // Bias  (ax,ay,az,gx,gy,gz)
+
+typedef struct {
```
Comment (dellaert) [inline examples/IMUKittiExampleGPS.cpp]: Prefer `struct ImuMeasurement {};`

File: examples/IMUKittiExampleGPS.cpp [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    double Time;
+    Vector3 Position; // x,y,z
+} gpsMeasurement_t;
+
+const string output_filename = "IMUKittyExampleGPSResults.csv";
+
+int main(int argc, char* argv[])
+{
+    string line;
```
Comment (dellaert) [inline examples/IMUKittiExampleGPS.cpp]: How about isolating reading of files in separate function?

File: examples/IMUKittiExampleGPS.cpp [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+    // Configure different variables
+    double tOffset = GPS_measurements[0].Time;
+    size_t firstGPSPose = 1;
+    size_t GPSskip = 10; // Skip this many GPS measurements each time
+    double g = 9.8;
+    auto w_coriolis = Vector3(); // zero vector
+
+    // Configure noise models
+    noiseModel::Diagonal::shared_ptr noiseModelGPS = noiseModel::Diagonal::Precisions((Vector(6) << Vector3::Constant(0), Vector3::Constant(1.0/0.07)).finished());
```
Comment (dellaert) [inline examples/IMUKittiExampleGPS.cpp]: auto

File: examples/IMUKittiExampleGPS.cpp [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+
+    // Set initial conditions for the estimated trajectory
+    auto currentPoseGlobal = Pose3(Rot3(), GPS_measurements[firstGPSPose].Position); // initial pose is the reference frame (navigation frame)
+    auto currentVelocityGlobal = Vector3(); // the vehicle is stationary at the beginning at position 0,0,0
+    auto currentBias = imuBias::ConstantBias(); // init with zero bias
+
+    noiseModel::Diagonal::shared_ptr sigma_init_x = noiseModel::Isotropic::Precisions((Vector(6) << Vector3::Constant(0), Vector3::Constant(1.0)).finished());
```
Comment (dellaert) [inline examples/IMUKittiExampleGPS.cpp]: auto

Also, (Vector(6) << Vector3::Constant(0), Vector3::Constant(1.0)).finished() is such a common case in this file and others maybe it's worth adding a small function at the top. PS It should be Diagonal ! This is actually crazy (and my fault) that you can do this: the constructor should be deleted from Isotropic.

File: examples/IMUKittiExampleGPS.cpp [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+    noiseModel::Diagonal::shared_ptr sigma_init_v = noiseModel::Isotropic::Sigma(3, 1000.0);
+    noiseModel::Diagonal::shared_ptr sigma_init_b = noiseModel::Isotropic::Sigmas((Vector(6) << Vector3::Constant(0.100), Vector3::Constant(5.00e-05)).finished());
+
+    // Set IMU preintegration parameters
+    Matrix33 measured_acc_cov = Matrix33::Identity(3,3) * pow(AccelerometerSigma,2);
```
Comment (dellaert) [inline examples/IMUKittiExampleGPS.cpp]: I_3x3

File: examples/IMUKittiExampleGPS.cpp [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+    NonlinearFactorGraph newFactors;
+    Values newValues; // values storing the initial estimates of new nodes in the factor graph
+
+    /// Main loop:
+    /// (1) we read the measurements
+    /// (2) we create the corresponding factors in the graph
+    /// (3) we solve the graph to obtain and optimal estimate of robot trajectory
+    printf("-- Starting main loop: inference is performed at each time step, but we plot trajectory every 10 steps\n");
+    size_t imuMeasurementIndex = 0;
+    for (size_t gpsMeasurementIndex = firstGPSPose; gpsMeasurementIndex < GPS_measurements.size() - 1; gpsMeasurementIndex++) {
```
Comment (dellaert) [inline examples/IMUKittiExampleGPS.cpp]: Question of style, but I prefer conciseness here. Suggest `gpsMeasurementIndex -> i`, `imuMeasurementIndex -> j`

File: examples/IMUKittiExampleGPS.cpp [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+        auto currentVelKey = V(gpsMeasurementIndex);
+        auto currentBiasKey = B(gpsMeasurementIndex);
+        double t = GPS_measurements[gpsMeasurementIndex].Time;
+
+        if (gpsMeasurementIndex == firstGPSPose) {
+            // Create initial estimate and prior on initial pose, velocity, and biases
+            newValues.insert(currentPoseKey, currentPoseGlobal);
+            newValues.insert(currentVelKey, currentVelocityGlobal);
+            newValues.insert(currentBiasKey, currentBias);
+            newFactors.add(PriorFactor<Pose3>(currentPoseKey, currentPoseGlobal, sigma_init_x));
```
Comment (dellaert) [inline examples/IMUKittiExampleGPS.cpp]: use emplace_shared?

File: examples/IMUKittiExampleGPS.cpp [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+
+    /// Main loop:
+    /// (1) we read the measurements
+    /// (2) we create the corresponding factors in the graph
+    /// (3) we solve the graph to obtain and optimal estimate of robot trajectory
+    printf("-- Starting main loop: inference is performed at each time step, but we plot trajectory every 10 steps\n");
+    size_t imuMeasurementIndex = 0;
+    for (size_t gpsMeasurementIndex = firstGPSPose; gpsMeasurementIndex < GPS_measurements.size() - 1; gpsMeasurementIndex++) {
+        // At each non=IMU measurement we initialize a new node in the graph
+        auto currentPoseKey = X(gpsMeasurementIndex);
```
Comment (dellaert) [inline examples/IMUKittiExampleGPS.cpp]: For conciseness and clarity (which comes from having to read less code), would just use X(i), V(i), B(i) in place below

File: examples/IMUKittiExampleGPS.cpp [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+            while (imuMeasurementIndex < IMU_measurements.size() && IMU_measurements[imuMeasurementIndex].Time <= t) {
+                if (IMU_measurements[imuMeasurementIndex].Time >= t_previous) {
+                    currentSummarizedMeasurement->integrateMeasurement(IMU_measurements[imuMeasurementIndex].Accelerometer, IMU_measurements[imuMeasurementIndex].Gyroscope, IMU_measurements[imuMeasurementIndex].dt);
+                    includedIMUmeasurementCount++;
+                }
+                imuMeasurementIndex++;
+            }
+
+            // Create IMU factor
+            auto previousPoseKey = X(gpsMeasurementIndex-1);
```
Comment (dellaert) [inline examples/IMUKittiExampleGPS.cpp]: same comment about keys in-pace -> X(i-1)

File: examples/IMUKittiExampleGPS.cpp [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+            auto previousVelKey = V(gpsMeasurementIndex-1);
+            auto previousBiasKey = B(gpsMeasurementIndex-1);
+
+            newFactors.add(ImuFactor(
+                                 previousPoseKey, previousVelKey,
+                                 currentPoseKey, currentVelKey,
+                                 previousBiasKey, *currentSummarizedMeasurement));
+
+            // Bias evolution as given in the IMU metadata
+            noiseModel::Diagonal::shared_ptr sigma_between_b = noiseModel::Diagonal::Sigmas((Vector(6) << Vector3::Constant(sqrt(includedIMUmeasurementCount) * AccelerometerBiasSigma), Vector3::Constant(sqrt(includedIMUmeasurementCount) * GyroscopeBiasSigma)).finished());
```
Comment (dellaert) [inline examples/IMUKittiExampleGPS.cpp]: auto

**Comment by varunagrawal** [comment]
@mindThomas any updates on this?

File: examples/IMUKittiExampleGPS.cpp (hunk 1 above)
Comment (nutonomy-thomas) [reply examples/IMUKittiExampleGPS.cpp:30]: Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

File: examples/IMUKittiExampleGPS.cpp (hunk 3 above)
Comment (nutonomy-thomas) [reply examples/IMUKittiExampleGPS.cpp:73]: Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

File: examples/IMUKittiExampleGPS.cpp (hunk 1 above)
Comment (nutonomy-thomas) [reply examples/IMUKittiExampleGPS.cpp:30]: Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

File: examples/IMUKittiExampleGPS.cpp (hunk 2 above)
Comment (nutonomy-thomas) [reply examples/IMUKittiExampleGPS.cpp]: Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

File: examples/IMUKittiExampleGPS.cpp (hunk 3 above)
Comment (nutonomy-thomas) [reply examples/IMUKittiExampleGPS.cpp:73]: Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

File: examples/IMUKittiExampleGPS.cpp (hunk 4 above)
Comment (nutonomy-thomas) [reply examples/IMUKittiExampleGPS.cpp]: Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

File: examples/IMUKittiExampleGPS.cpp (hunk 5 above)
Comment (nutonomy-thomas) [reply examples/IMUKittiExampleGPS.cpp]: I have addressed the Isotropic to Diagonal part in b48bdfdc696328be3fedfe11db18b83d3341e7bc
The Vector(6) remains the same to be explicit.

File: examples/IMUKittiExampleGPS.cpp (hunk 7 above)
Comment (nutonomy-thomas) [reply examples/IMUKittiExampleGPS.cpp]: Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

File: examples/IMUKittiExampleGPS.cpp (hunk 8 above)
Comment (nutonomy-thomas) [reply examples/IMUKittiExampleGPS.cpp]: Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

File: examples/IMUKittiExampleGPS.cpp (hunk 9 above)
Comment (nutonomy-thomas) [reply examples/IMUKittiExampleGPS.cpp]: Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

File: examples/IMUKittiExampleGPS.cpp (hunk 10 above)
Comment (nutonomy-thomas) [reply examples/IMUKittiExampleGPS.cpp]: Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

File: examples/IMUKittiExampleGPS.cpp (hunk 11 above)
Comment (nutonomy-thomas) [reply examples/IMUKittiExampleGPS.cpp]: Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

**Comment by mindThomas** [comment]
I have addressed your review comments and revised the formatting to conform to the Google C++ style.

File: examples/IMUKittiExampleGPS.cpp (hunk 11 above)
Comment (mindThomas) [reply examples/IMUKittiExampleGPS.cpp]: Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

File: examples/IMUKittiExampleGPS.cpp (hunk 10 above)
Comment (mindThomas) [reply examples/IMUKittiExampleGPS.cpp]: Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

File: examples/IMUKittiExampleGPS.cpp (hunk 5 above)
Comment (mindThomas) [reply examples/IMUKittiExampleGPS.cpp]: I have addressed the Isotropic to Diagonal part in b48bdfdc696328be3fedfe11db18b83d3341e7bc
The Vector(6) remains the same to be explicit.

File: examples/IMUKittiExampleGPS.cpp (hunk 7 above)
Comment (mindThomas) [reply examples/IMUKittiExampleGPS.cpp]: Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

File: examples/IMUKittiExampleGPS.cpp (hunk 8 above)
Comment (mindThomas) [reply examples/IMUKittiExampleGPS.cpp]: Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

File: examples/IMUKittiExampleGPS.cpp (hunk 9 above)
Comment (mindThomas) [reply examples/IMUKittiExampleGPS.cpp]: Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

File: examples/IMUKittiExampleGPS.cpp (hunk 4 above)
Comment (mindThomas) [reply examples/IMUKittiExampleGPS.cpp]: Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

File: examples/IMUKittiExampleGPS.cpp (hunk 3 above)
Comment (mindThomas) [reply examples/IMUKittiExampleGPS.cpp:73]: Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

File: examples/IMUKittiExampleGPS.cpp (hunk 1 above)
Comment (mindThomas) [reply examples/IMUKittiExampleGPS.cpp:30]: Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

File: examples/IMUKittiExampleGPS.cpp (hunk 2 above)
Comment (mindThomas) [reply examples/IMUKittiExampleGPS.cpp]: Addressed in b48bdfdc696328be3fedfe11db18b83d3341e7bc

**Comment by varunagrawal** [comment]
@mindThomas is there a reason your code uses the convention `Kitty` rather than `Kitti`?

**Comment by mindThomas** [comment]
@varunagrawal That was a brain fart from my side. This has been corrected now.

**Pull request merged**